### PR TITLE
storage: AUCTION load generator

### PR DESCRIPTION
**Pull request by mjibson**
This also adds `CREATE VIEWS` support for supporting load generators.

### Motivation

  * This PR adds a feature that has not yet been specified.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

File: src/storage/src/types/sources.rs
Comment (mjibson) [inline src/storage/src/types/sources.rs:1592]: This is my new idea for how generators should work: each invocation can produce many rows, that way the generator can make related rows if needed, then the source accesses the generator as an iterator.

File: src/sql/src/plan/statement/ddl.rs
Comment (mjibson) [inline src/sql/src/plan/statement/ddl.rs:1544]: At one point I had the new `views()` fn return a `RelationDesc` instead of the `(string, string)` it currently does, but I couldn't find a way to convert `ScalarType` into a resolved data type (or into a string for resolving with `names::resolve`) needed by `Expr::Cast`.

File: src/storage/src/types/sources.rs
Comment (sjwiesman) [reply src/storage/src/types/sources.rs:1592]: Will all the rows be relocked at the same timestamp? I'm wondering if this can be used to show off transactional results?

File: src/sql/src/plan/statement/ddl.rs
Comment (sjwiesman) [reply src/sql/src/plan/statement/ddl.rs:1544]: I'll let someone make the final call, but my take is that this isn't Kafka. No one is using this thing outside a getting started experience and being pragmatic is a resonable take.

**Review by sjwiesman — APPROVED** [review]
Thank you so much for doing this so quickly! Having a demo source like this will go so far in improving our getting started experience. There is nothing I see here as blocking this PR, but I also don't spend a lot of time in this repo, I wouldn't be offended if someone else wanted changes before merging :)

File: src/storage/src/types/sources.rs
Comment (mjibson) [reply src/storage/src/types/sources.rs:1557]: This is what my question above was about. I couldn't find a way to make these not strings.

File: src/storage/src/types/sources.rs
Comment (mjibson) [reply src/storage/src/types/sources.rs:1592]: They currently do not, but we could change them to be reclocked to the same timestamp if some load generator needed that, yes.

Comment (sjwiesman) [reply src/storage/src/types/sources.rs:1592]: I wouldn’t block this PR on that change, I’m just trying to imagine the possibilities

File: src/storage/src/types/sources.rs
Comment (sjwiesman) [reply src/storage/src/types/sources.rs:1557]: Oh, gotcha

File: src/sql/src/plan/statement/ddl.rs
Comment (benesch) [reply src/sql/src/plan/statement/ddl.rs:1544]: You can take a `ScalarType` to a `pgrepr::Type` and then take _that_ to a string. We're already doing the second half of this in `CREATE VIEWS` for PostgreSQL; would be good to wrap it up in a function and use it in both places!

https://github.com/MaterializeInc/materialize/blob/625c4888772d71a10c14b6b1f4e90eb24203a028/src/sql/src/plan/statement/ddl.rs#L1592-L1606

File: src/storage/src/types/sources.rs
Comment (benesch) [inline src/storage/src/types/sources.rs:1622]: This just happens to produce a parseable representation, right? `strconv::format_timestamptz` seems like a safer choice.

File: src/storage/src/types/sources.rs
Comment (benesch) [inline src/storage/src/types/sources.rs:1643]: Since it's likely that we'll have many of these load generators of increasing complexity, I'd love to plan for that now and have a separate module for each load generator source. This file is already too long!

File: src/storage/src/types/sources.rs
Comment (benesch) [reply src/storage/src/types/sources.rs:1622]: Bummer that you have to roundtrip through the string representation at all. Wish we had a proper union type.

File: src/storage/src/types/sources.rs
Comment (sjwiesman) [reply src/storage/src/types/sources.rs:1643]: +1. What would be really great is a single trait, so I or devex could be more involved

Comment (benesch) [reply src/storage/src/types/sources.rs:1643]: Unfortunately we've tried and failed to make that trait for several years. 🙈

**Comment by mjibson** [comment]
Have made requested changes. Big code movement to try to make each generator its own file. Is this about what you meant? I really don't like some of it, especially the `as_generator` fn, but not better way reveals itself.

**Review by benesch — APPROVED** [review]
This looks great! Thanks very much for the code restructuring. If you wanted to avoid `as_generator`, you could `impl Generator for LoadGenerator`, either by hand or using something like [`enum_dispatch`](https://docs.rs/enum_dispatch/latest/enum_dispatch/index.html). But `as_generator` seems totally fine to me, FWIW!

**Pull request merged**
